src/Functions/IFunction.cpp

```diff
@@ -188,17 +202,101 @@ ColumnPtr IExecutableFunction::defaultImplementationForNulls(
                 "is expected to return Nullable result, got {}",
                 getName(),
                 result_type->getName());
+    }
 
+    if (null_presence.has_null_constant)
+    {
+        /// If any of the input arguments is null literal, the result is null constant.
         return result_type->createColumnConstWithDefaultValue(input_rows_count);
     }
 
     if (null_presence.has_nullable)
     {
+        /// Usually happens during analyzing. We should return non-const column to avoid wrong constant folding.
```

Comment:
Uninitiated developers will find below code quite difficult. It has comments about the "how" but not the "why". Could you please add more context and motivation from the PR message of https://github.com/ClickHouse/ClickHouse/pull/60129 ?